blocks/main_loop.py

```diff
@@ -42,6 +45,90 @@
 a model. Check documentation of your extensions."""
 
 
+class Timer(object):
+    def __init__(self):
+        self.total = defaultdict(int)
+        self.current = []
+        self.order = OrderedSet()
+
+    def enter(self, name):
+        self.current.append(name)
+        # We record the order in which sections were first called
+        self.order.add(tuple(self.current))
```

Comment:
Main issue with `OrderedDict` is that it's nearly twice as slow as `OrderedSet`, which can become significant if the user is performing batch-monitoring on a fast model (i.e. when `enter` gets called a lot). I'd have to do some benchmarking to see if that difference is noticeable.